rust_src/src/buffers.rs

```diff
@@ -105,6 +110,90 @@ pub type LispBufferRef = ExternalPtr<Lisp_Buffer>;
 pub type LispOverlayRef = ExternalPtr<Lisp_Overlay>;
 
 impl LispBufferRef {
+    pub fn create_new(name: LispStringRef) -> LispBufferRef {
+        if name.is_empty() {
+            error!("Empty string for buffer name is not allowed");
+        }
+
+        let mut b = unsafe { ExternalPtr::new(allocate_buffer()) };
+        // An ordinary b uses its own b_text
+        b.text = &mut b.own_text;
+        b.base_buffer = ptr::null_mut();
+        // No one shares the text with us.
+        b.indirections = 0;
+        // No one shows us now.
+        b.window_count = 0;
+
+        let gap_size = 20;
+        let mut b_text = unsafe { &mut *b.text };
+        b_text.gap_size = gap_size;
+        unsafe {
+            block_input();
+            alloc_buffer_text(b.as_mut(), gap_size + 1);
+            unblock_input();
+            if b.beg_addr().is_null() {
+                buffer_memory_full(gap_size + 1);
+            }
+        }
+
+        b.set_pt_both(b.beg(), b.beg_byte());
+        b.set_begv_both(b.beg(), b.beg_byte());
+        b.set_zv_both(b.beg(), b.beg_byte());
+        b_text.gpt = b.beg();
+        b_text.gpt_byte = b.beg_byte();
+        b_text.z = b.beg();
+        b_text.z_byte = b.beg_byte();
+        b_text.modiff = 1;
+        b_text.chars_modiff = 1;
+        b_text.overlay_modiff = 1;
+        b_text.save_modiff = 1;
+        b_text.compact = 1;
+        b_text.intervals = ptr::null_mut();
+        b_text.unchanged_modified = 1;
+        b_text.overlay_unchanged_modified = 1;
+        b_text.end_unchanged = 0;
+        b_text.beg_unchanged = 0;
+        unsafe {
+            // Put an anchor '\0'.
+            *b.z_addr() = 0;
+            *b.gap_start_addr() = 0;
+        }
+        b_text.set_inhibit_shrinking(false);
+        b_text.set_redisplay(false);
+
+        b.newline_cache = ptr::null_mut();
+        b.width_run_cache = ptr::null_mut();
+        b.bidi_paragraph_cache = ptr::null_mut();
+        b.width_table_ = Qnil;
+        b.set_prevent_redisplay_optimizations_p(true);
+
+        // An ordinary buffer normally doesn't need markers to handle BEGV and ZV.
+        b.pt_marker_ = Qnil;
+        b.begv_marker_ = Qnil;
+        b.zv_marker_ = Qnil;
+
+        let mut name: LispStringRef = unsafe { Fcopy_sequence(name.into()) }.force_string();
+        name.set_intervals(ptr::null_mut());
+        b.name_ = name.into();
+        b.undo_list_ = if name.byte_at(0) != b' ' { Qnil } else { Qt };
```

Comment:
Would adding a `bool_as_lisp` function help make our checks read better?

`b.undo_list_ = bool_as_lisp(name.byte_at(0) == b' ');`

Reply:
may be add a trait with method `as_lisp` which in turn calls `into` ??

Reply:
My point was that sometimes the explicit naming makes the code easier to read. `as_lisp` is no better than `into` really.